# Practica de CNN

## Grupo:
* Dussan Freire
* Jhonny Camacho

### Links Utilizados
* __https://keras.io/examples/vision/mnist_convnet/__
* __https://victorzhou.com/blog/keras-cnn-tutorial/__

## Librerias usadas

In [2]:
from torchvision.datasets import MNIST
from torchvision import transforms
import numpy as np
#import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Dense, Flatten
from tensorflow.keras.utils import to_categorical
import tensorflow as tf
from tensorflow import keras
import timeit
from datetime import datetime
from tensorflow.keras import layers
from tensorflow.keras.layers import Dropout
from tensorflow.keras.layers import Dense

In [ ]:
# !pip install tensorflow numpy mnist

# Extracción datasets

In [3]:
import numpy as np
from tensorflow import keras
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Dense, Flatten
from tensorflow.keras.utils import to_categorical

# The first time you run this might be a bit slow, since the
# mnist package has to download and cache the data.
mnist = keras.datasets.mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

print("X_train shape", train_images.shape)
print("y_train shape", train_labels.shape)
print("X_test shape", test_images.shape)
print("y_test shape", test_labels.shape)

11493376/11490434 [==============================] - 0s 0us/step
X_train shape (60000, 28, 28)
y_train shape (60000,)
X_test shape (10000, 28, 28)
y_test shape (10000,)


# Experimentos

## Set Up para los experimentos

In [4]:
# Normalize the images.
train_images = (train_images / 255) - 0.5
test_images = (test_images / 255) - 0.5

# Reshape the images.
train_images = np.expand_dims(train_images, axis=3)
test_images = np.expand_dims(test_images, axis=3)

## Experimento 1
Se aumento la cantidad de filtros de 8 a 15

### Caracteristicas:  

In [5]:
# parametros utilizados
num_filters = 15
filter_size = 3
pool_size = 2
epochs = 3

### Desarrollo

In [6]:
# Contruccion del modelo
model = Sequential([
  Conv2D(num_filters, filter_size, input_shape=(28, 28, 1)),
  MaxPooling2D(pool_size=pool_size),
  Flatten(),
  Dense(10, activation='softmax'),
])

# Compilacion del modelo
model.compile(
  'adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Crear una closure para poder medir el tiempo
def run_experiment1():
      model.fit(
      train_images,
      to_categorical(train_labels),
      epochs=epochs,
      validation_data=(test_images, to_categorical(test_labels)))

# Entrenar el modela
execution_time = timeit.timeit(run_experiment1, number=1)

print("tiempo_de_entrenamiento: ", execution_time,"[s]")

Epoch 1/3
1875/1875 [==============================] - 22s 11ms/step - loss: 0.3134 - accuracy: 0.9089 - val_loss: 0.1628 - val_accuracy: 0.9541
Epoch 2/3
1875/1875 [==============================] - 21s 11ms/step - loss: 0.1374 - accuracy: 0.9603 - val_loss: 0.1065 - val_accuracy: 0.9695
Epoch 3/3
1875/1875 [==============================] - 21s 11ms/step - loss: 0.0965 - accuracy: 0.9722 - val_loss: 0.0854 - val_accuracy: 0.9741
tiempo_de_entrenamiento:  82.67568148299998 [s]


### Guardamos los pesos

In [ ]:
model.save_weights('experimeto1.h5')

### Test

In [8]:
# Predict on the first 5 test images.
predictions = model.predict(test_images[:5])

# Print our model's predictions.
print(np.argmax(predictions, axis=1)) 

# Check our predictions against the ground truths.
print(test_labels[:5]) 
# print(test_labels[:5]) # [7, 2, 1, 0, 4]

[7 2 1 0 4]
[7 2 1 0 4]


### Concluision
Se mejoro el resultado y pero tomo un poco mas de tiempo entrenar.

# Experimento numero 2
Se aumento tamano del filro

### Caracteriscidas:

In [9]:
# parametros utilizados
num_filters = 8
filter_size = 10
pool_size = 2
epochs = 3

### Desarrollo

In [10]:
# Contruccion del modelo
model = Sequential([
  Conv2D(num_filters, filter_size, input_shape=(28, 28, 1)),
  MaxPooling2D(pool_size=pool_size),
  Flatten(),
  Dense(10, activation='softmax'),
])

# Compilacion del modelo
model.compile(
  'adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Crear una closure para poder medir el tiempo
def run_experiment1():
      model.fit(
      train_images,
      to_categorical(train_labels),
      epochs=epochs,
      validation_data=(test_images, to_categorical(test_labels)))

# Entrenar el modela
execution_time = timeit.timeit(run_experiment1, number=1)

print("tiempo_de_entrenamiento: ", execution_time,"[s]")

Epoch 1/3
1875/1875 [==============================] - 28s 15ms/step - loss: 0.3130 - accuracy: 0.9102 - val_loss: 0.1570 - val_accuracy: 0.9557
Epoch 2/3
1875/1875 [==============================] - 27s 14ms/step - loss: 0.1360 - accuracy: 0.9606 - val_loss: 0.1007 - val_accuracy: 0.9707
Epoch 3/3
1875/1875 [==============================] - 27s 14ms/step - loss: 0.0965 - accuracy: 0.9716 - val_loss: 0.0797 - val_accuracy: 0.9755
tiempo_de_entrenamiento:  82.39007004299998 [s]


### Guardamos los pesos

In [ ]:
model.save_weights('experimeto2.h5')

### Test

In [11]:
# Predict on the first 5 test images.
predictions = model.predict(test_images[:5])

# Print our model's predictions.
print(np.argmax(predictions, axis=1)) 

# Check our predictions against the ground truths.
print(test_labels[:5]) 
# print(test_labels[:5]) # [7, 2, 1, 0, 4]

[7 2 1 0 4]
[7 2 1 0 4]


### Concluision
La exatitud de la ultima capa 0.9716 en un tiempo de 82.39 segundos

# Experimento numero 3
Se agrando el pool size de 2 a 4

### Caracteriscidas:

In [13]:
# parametros utilizados
num_filters = 8
filter_size = 3
pool_size = 4
epochs = 3

### Desarrollo

In [14]:
# Contruccion del modelo
model = Sequential([
  Conv2D(num_filters, filter_size, input_shape=(28, 28, 1)),
  MaxPooling2D(pool_size=pool_size),
  Flatten(),
  Dense(10, activation='softmax'),
])

# Compilacion del modelo
model.compile(
  'adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Crear una closure para poder medir el tiempo
def run_experiment1():
      model.fit(
      train_images,
      to_categorical(train_labels),
      epochs=epochs,
      validation_data=(test_images, to_categorical(test_labels)))

# Entrenar el modela
execution_time = timeit.timeit(run_experiment1, number=1)

print("tiempo_de_entrenamiento: ", execution_time,"[s]")

Epoch 1/3
1875/1875 [==============================] - 16s 8ms/step - loss: 0.4046 - accuracy: 0.8854 - val_loss: 0.1764 - val_accuracy: 0.9513
Epoch 2/3
1875/1875 [==============================] - 15s 8ms/step - loss: 0.1534 - accuracy: 0.9550 - val_loss: 0.1189 - val_accuracy: 0.9633
Epoch 3/3
1875/1875 [==============================] - 16s 8ms/step - loss: 0.1172 - accuracy: 0.9649 - val_loss: 0.0957 - val_accuracy: 0.9703
tiempo_de_entrenamiento:  46.62441658500006 [s]


### Guardamos los pesos

In [ ]:
model.save_weights('experimeto3.h5')

### Test

In [15]:
# Predict on the first 5 test images.
predictions = model.predict(test_images[:5])

# Print our model's predictions.
print(np.argmax(predictions, axis=1)) 

# Check our predictions against the ground truths.
print(test_labels[:5]) 
# print(test_labels[:5]) # [7, 2, 1, 0, 4]

[7 2 1 0 4]
[7 2 1 0 4]


### Concluision
La exatitud de la ultima capa 0.9649 en un tiempo de 46.62 segundos

# Experimento numero 4
Se aumentaron las epocas de 3 a 6

### Caracteriscidas:

In [16]:
# parametros utilizados
num_filters = 8
filter_size = 3
pool_size = 2
epochs = 6

### Desarrollo

In [17]:
# Contruccion del modelo
model = Sequential([
  Conv2D(num_filters, filter_size, input_shape=(28, 28, 1)),
  MaxPooling2D(pool_size=pool_size),
  Flatten(),
  Dense(10, activation='softmax'),
])

# Compilacion del modelo
model.compile(
  'adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Crear una closure para poder medir el tiempo
def run_experiment1():
      model.fit(
      train_images,
      to_categorical(train_labels),
      epochs=epochs,
      validation_data=(test_images, to_categorical(test_labels)))

# Entrenar el modela
execution_time = timeit.timeit(run_experiment1, number=1)

print("tiempo_de_entrenamiento: ", execution_time,"[s]")

Epoch 1/6
1875/1875 [==============================] - 17s 9ms/step - loss: 0.3553 - accuracy: 0.8987 - val_loss: 0.2037 - val_accuracy: 0.9404
Epoch 2/6
1875/1875 [==============================] - 17s 9ms/step - loss: 0.1697 - accuracy: 0.9524 - val_loss: 0.1316 - val_accuracy: 0.9612
Epoch 3/6
1875/1875 [==============================] - 17s 9ms/step - loss: 0.1188 - accuracy: 0.9665 - val_loss: 0.1065 - val_accuracy: 0.9694
Epoch 4/6
1875/1875 [==============================] - 16s 8ms/step - loss: 0.0963 - accuracy: 0.9717 - val_loss: 0.0935 - val_accuracy: 0.9708
Epoch 5/6
1875/1875 [==============================] - 16s 9ms/step - loss: 0.0839 - accuracy: 0.9758 - val_loss: 0.0843 - val_accuracy: 0.9738
Epoch 6/6
1875/1875 [==============================] - 17s 9ms/step - loss: 0.0745 - accuracy: 0.9777 - val_loss: 0.0830 - val_accuracy: 0.9721
tiempo_de_entrenamiento:  142.37407452899993 [s]


In [18]:
model.save_weights('experimeto4.h5')

### Test

In [ ]:
# Predict on the first 5 test images.
predictions = model.predict(test_images[:5])

# Print our model's predictions.
print(np.argmax(predictions, axis=1)) 

# Check our predictions against the ground truths.
print(test_labels[:5].numpy().tolist()) 
# print(test_labels[:5]) # [7, 2, 1, 0, 4]

[5 0 4 1 9]
[5, 0, 4, 1, 9]


### Concluision
La exatitud de la ultima capa 0.9777 en un tiempo de 142.37 segundos

# Experimento  5
Se aumento una capa de conv2d de 1 a 2

### Caracteriscidas:

In [19]:
# parametros utilizados
num_filters = 8
filter_size = 3
pool_size = 2
epochs = 6

### Desarrollo

In [20]:
# Contruccion del modelo
model = Sequential([
  Conv2D(num_filters, filter_size, input_shape=(28, 28, 1)),
  Conv2D(num_filters, filter_size, input_shape=(28, 28, 1)),
  MaxPooling2D(pool_size=pool_size),
  Flatten(),
  Dense(10, activation='softmax'),
])

# Compilacion del modelo
model.compile(
  'adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Crear una closure para poder medir el tiempo
def run_experiment1():
      model.fit(
      train_images,
      to_categorical(train_labels),
      epochs=epochs,
      validation_data=(test_images, to_categorical(test_labels)))

# Entrenar el modela
execution_time = timeit.timeit(run_experiment1, number=1)

print("tiempo_de_entrenamiento: ", execution_time,"[s]")

Epoch 1/6
1875/1875 [==============================] - 34s 18ms/step - loss: 0.2602 - accuracy: 0.9232 - val_loss: 0.1148 - val_accuracy: 0.9655
Epoch 2/6
1875/1875 [==============================] - 35s 18ms/step - loss: 0.1001 - accuracy: 0.9706 - val_loss: 0.0894 - val_accuracy: 0.9712
Epoch 3/6
1875/1875 [==============================] - 35s 19ms/step - loss: 0.0790 - accuracy: 0.9759 - val_loss: 0.0766 - val_accuracy: 0.9762
Epoch 4/6
1875/1875 [==============================] - 34s 18ms/step - loss: 0.0669 - accuracy: 0.9791 - val_loss: 0.0715 - val_accuracy: 0.9787
Epoch 5/6
1875/1875 [==============================] - 33s 18ms/step - loss: 0.0599 - accuracy: 0.9811 - val_loss: 0.0754 - val_accuracy: 0.9763
Epoch 6/6
1875/1875 [==============================] - 34s 18ms/step - loss: 0.0550 - accuracy: 0.9827 - val_loss: 0.0679 - val_accuracy: 0.9791
tiempo_de_entrenamiento:  262.3647253710001 [s]


In [21]:
model.save_weights('experimeto5.h5')

### Test

In [22]:
# Predict on the first 5 test images.
predictions = model.predict(test_images[:5])

# Print our model's predictions.
print(np.argmax(predictions, axis=1)) 

# Check our predictions against the ground truths.
print(test_labels[:5]) 
# print(test_labels[:5]) # [7, 2, 1, 0, 4]

[7 2 1 0 4]
[7 2 1 0 4]


### Concluision
La exatitud de la ultima capa 0.9827 en un tiempo de 262.36 segundos

# Experimento 6
Se aumento una capa "dropout", para evitar overfitting

### Caracteriscidas:

In [23]:
# parametros utilizados
num_filters = 8
filter_size = 3
pool_size = 2
epochs = 6

### Desarrollo

In [24]:
# Contruccion del modelo
model = Sequential([
  Conv2D(num_filters, filter_size, input_shape=(28, 28, 1)),
  MaxPooling2D(pool_size=pool_size),
  Dropout(0.5),
  Flatten(),
  Dense(10, activation='softmax'),
])

# Compilacion del modelo
model.compile(
  'adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Crear una closure para poder medir el tiempo
def run_experiment1():
      model.fit(
      train_images,
      to_categorical(train_labels),
      epochs=epochs,
      validation_data=(test_images, to_categorical(test_labels)))

# Entrenar el modela
execution_time = timeit.timeit(run_experiment1, number=1)

print("tiempo_de_entrenamiento: ", execution_time,"[s]")

Epoch 1/6
1875/1875 [==============================] - 21s 11ms/step - loss: 0.4357 - accuracy: 0.8701 - val_loss: 0.2162 - val_accuracy: 0.9381
Epoch 2/6
1875/1875 [==============================] - 21s 11ms/step - loss: 0.2632 - accuracy: 0.9206 - val_loss: 0.1668 - val_accuracy: 0.9520
Epoch 3/6
1875/1875 [==============================] - 24s 13ms/step - loss: 0.2288 - accuracy: 0.9315 - val_loss: 0.1463 - val_accuracy: 0.9584
Epoch 4/6
1875/1875 [==============================] - 28s 15ms/step - loss: 0.2140 - accuracy: 0.9350 - val_loss: 0.1333 - val_accuracy: 0.9609
Epoch 5/6
1875/1875 [==============================] - 25s 13ms/step - loss: 0.2002 - accuracy: 0.9390 - val_loss: 0.1222 - val_accuracy: 0.9632
Epoch 6/6
1875/1875 [==============================] - 20s 10ms/step - loss: 0.1957 - accuracy: 0.9405 - val_loss: 0.1198 - val_accuracy: 0.9648
tiempo_de_entrenamiento:  142.38013125399993 [s]


In [25]:
model.save_weights('experimeto6.h5')

### Test

In [26]:
# Predict on the first 5 test images.
predictions = model.predict(test_images[:5])

# Print our model's predictions.
print(np.argmax(predictions, axis=1)) 

# Check our predictions against the ground truths.
print(test_labels[:5]) 
# print(test_labels[:5]) # [7, 2, 1, 0, 4]

[7 2 1 0 4]
[7 2 1 0 4]


### Concluision
La exatitud de la ultima capa 0.9405 en un tiempo de 142.380 segundos

# Experimento  7
Se aumento una capa anterior a la capa de salida de 64 unidades de activacion utilizando la funcion relu

### Caracteriscidas:

In [27]:
# parametros utilizados
num_filters = 8
filter_size = 3
pool_size = 2
epochs = 6

### Desarrollo

In [28]:
# Contruccion del modelo
model = Sequential([
  Conv2D(num_filters, filter_size, input_shape=(28, 28, 1)),
  MaxPooling2D(pool_size=pool_size),
  Flatten(),
  Dense(64, activation='relu'),
  Dense(10, activation='softmax'),
])

# Compilacion del modelo
model.compile(
  'adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Crear una closure para poder medir el tiempo
def run_experiment1():
      model.fit(
      train_images,
      to_categorical(train_labels),
      epochs=epochs,
      validation_data=(test_images, to_categorical(test_labels)))

# Entrenar el modela
execution_time = timeit.timeit(run_experiment1, number=1)

print("tiempo_de_entrenamiento: ", execution_time,"[s]")

Epoch 1/6
1875/1875 [==============================] - 20s 10ms/step - loss: 0.2688 - accuracy: 0.9221 - val_loss: 0.1215 - val_accuracy: 0.9625
Epoch 2/6
1875/1875 [==============================] - 19s 10ms/step - loss: 0.1006 - accuracy: 0.9696 - val_loss: 0.0805 - val_accuracy: 0.9741
Epoch 3/6
1875/1875 [==============================] - 19s 10ms/step - loss: 0.0696 - accuracy: 0.9787 - val_loss: 0.0617 - val_accuracy: 0.9798
Epoch 4/6
1875/1875 [==============================] - 19s 10ms/step - loss: 0.0536 - accuracy: 0.9836 - val_loss: 0.0601 - val_accuracy: 0.9810
Epoch 5/6
1875/1875 [==============================] - 19s 10ms/step - loss: 0.0424 - accuracy: 0.9866 - val_loss: 0.0656 - val_accuracy: 0.9798
Epoch 6/6
1875/1875 [==============================] - 19s 10ms/step - loss: 0.0328 - accuracy: 0.9895 - val_loss: 0.0683 - val_accuracy: 0.9807
tiempo_de_entrenamiento:  142.43029125600015 [s]


In [29]:
model.save_weights('experimeto7.h5')

### Test

In [30]:
# Predict on the first 5 test images.
predictions = model.predict(test_images[:5])

# Print our model's predictions.
print(np.argmax(predictions, axis=1)) 

# Check our predictions against the ground truths.
print(test_labels[:5]) 
# print(test_labels[:5]) # [7, 2, 1, 0, 4]

[7 2 1 0 4]
[7 2 1 0 4]


### Concluision
La exatitud de la ultima capa 0.9895 en un tiempo de 142.430 segundos

# Experimento  8
Se jugo con los parametros de la capa de convulucion

### Caracteriscidas:

In [31]:
# parametros utilizados
num_filters = 8
filter_size = 3
pool_size = 2
epochs = 6

### Desarrollo

In [32]:
# Contruccion del modelo
model = Sequential([
  Conv2D(
    num_filters,
    filter_size,
    input_shape=(28, 28, 1),
    strides=2,
    padding='same',
    activation='relu',
  ),
  MaxPooling2D(pool_size=pool_size),
  Flatten(),
  Dense(10, activation='softmax'),
])

# Compilacion del modelo
model.compile(
  'adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Crear una closure para poder medir el tiempo
def run_experiment1():
      model.fit(
      train_images,
      to_categorical(train_labels),
      epochs=epochs,
      validation_data=(test_images, to_categorical(test_labels)))

# Entrenar el modela
execution_time = timeit.timeit(run_experiment1, number=1)

print("tiempo_de_entrenamiento: ", execution_time,"[s]")

Epoch 1/6
1875/1875 [==============================] - 8s 4ms/step - loss: 0.4804 - accuracy: 0.8723 - val_loss: 0.2253 - val_accuracy: 0.9325
Epoch 2/6
1875/1875 [==============================] - 8s 4ms/step - loss: 0.1967 - accuracy: 0.9428 - val_loss: 0.1645 - val_accuracy: 0.9493
Epoch 3/6
1875/1875 [==============================] - 8s 4ms/step - loss: 0.1607 - accuracy: 0.9520 - val_loss: 0.1421 - val_accuracy: 0.9552
Epoch 4/6
1875/1875 [==============================] - 8s 4ms/step - loss: 0.1420 - accuracy: 0.9576 - val_loss: 0.1267 - val_accuracy: 0.9600
Epoch 5/6
1875/1875 [==============================] - 8s 4ms/step - loss: 0.1309 - accuracy: 0.9601 - val_loss: 0.1280 - val_accuracy: 0.9595
Epoch 6/6
1875/1875 [==============================] - 8s 4ms/step - loss: 0.1210 - accuracy: 0.9636 - val_loss: 0.1136 - val_accuracy: 0.9630
tiempo_de_entrenamiento:  82.3931419280002 [s]


In [33]:
model.save_weights('experimeto8.h5')

### Test

In [34]:
# Predict on the first 5 test images.
predictions = model.predict(test_images[:5])

# Print our model's predictions.
print(np.argmax(predictions, axis=1)) 

# Check our predictions against the ground truths.
print(test_labels[:5]) 
# print(test_labels[:5]) # [7, 2, 1, 0, 4]

[7 2 1 0 4]
[7 2 1 0 4]


### Concluision
La exatitud de la ultima capa 0.9636 en un tiempo de 82.3931 segundos

# Experimento sacado del link:
__https://keras.io/examples/vision/mnist_convnet/__


### Caracteristicas

In [ ]:
# Parametros del modelo 
num_classes = 10
input_shape = (28, 28, 1)
batch_size = 128
epochs = 15

### Desarrollo

In [ ]:
x_train = train_set.train_data
y_train = train_set.train_labels
x_test  = train_set.test_data
y_test = train_set.test_labels


# Scale images to the [0, 1] range
x_train = x_train / 255
x_test = x_test / 255

# Make sure images have shape (28, 28, 1)
x_train = np.expand_dims(x_train, -1)
x_test = np.expand_dims(x_test, -1)
print("x_train shape:", x_train.shape)
print(x_train.shape[0], "train samples")
print(x_test.shape[0], "test samples")

# Convertir a matrices binarias
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)
# Contruccion del modelo
model = keras.Sequential(
    [
        keras.Input(shape=input_shape),
        layers.Conv2D(32, kernel_size=(3, 3), activation="relu"),
        layers.MaxPooling2D(pool_size=(2, 2)),
        layers.Conv2D(64, kernel_size=(3, 3), activation="relu"),
        layers.MaxPooling2D(pool_size=(2, 2)),
        layers.Flatten(),
        layers.Dropout(0.5),
        layers.Dense(num_classes, activation="softmax"),
    ]
)

x_train shape: (60000, 28, 28, 1)
60000 train samples
60000 test samples


In [ ]:
model.summary()

Model: "sequential_8"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_9 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_8 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_10 (Conv2D)           (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_9 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
flatten_8 (Flatten)          (None, 1600)              0         
_________________________________________________________________
dropout_1 (Dropout)          (None, 1600)              0         
_________________________________________________________________
dense_9 (Dense)              (None, 10)               

### Modelo de entrenamiento

In [ ]:
model.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"])
model.fit(x_train, y_train, batch_size=batch_size, epochs=epochs, validation_split=0.1)
# Evaluacion del modelo entrenado
score = model.evaluate(x_test, y_test, verbose=0)
print("Test loss:", score[0])
print("Test accuracy:", score[1])

Epoch 1/15
422/422 [==============================] - 42s 99ms/step - loss: 0.3628 - accuracy: 0.8897 - val_loss: 0.0816 - val_accuracy: 0.9773
Epoch 2/15
422/422 [==============================] - 41s 98ms/step - loss: 0.1138 - accuracy: 0.9650 - val_loss: 0.0594 - val_accuracy: 0.9832
Epoch 3/15
422/422 [==============================] - 37s 89ms/step - loss: 0.0881 - accuracy: 0.9729 - val_loss: 0.0484 - val_accuracy: 0.9860
Epoch 4/15
422/422 [==============================] - 36s 86ms/step - loss: 0.0728 - accuracy: 0.9773 - val_loss: 0.0437 - val_accuracy: 0.9882
Epoch 5/15
422/422 [==============================] - 36s 85ms/step - loss: 0.0659 - accuracy: 0.9792 - val_loss: 0.0417 - val_accuracy: 0.9888
Epoch 6/15
422/422 [==============================] - 36s 85ms/step - loss: 0.0581 - accuracy: 0.9818 - val_loss: 0.0374 - val_accuracy: 0.9900
Epoch 7/15
422/422 [==============================] - 37s 88ms/step - loss: 0.0531 - accuracy: 0.9825 - val_loss: 0.0367 - val_accuracy:

### Evaluate the trained model

In [ ]:
score = model.evaluate(x_test, y_test, verbose=0)
print("Test loss:", score[0])
print("Test accuracy:", score[1])

Test loss: 0.01687825657427311
Test accuracy: 0.9950500130653381
